### PR TITLE
HDDS-5082. Create unit (!) test for OzoneClient

### PR DESCRIPTION
**Pull request by elek**
See: https://issues.apache.org/jira/browse/HDDS-5082

## What changes were proposed in this pull request?

Today `OzoneClient` is tested only by integration tests (`TestOzoneRpcClientAbstract` from integration-test and all the sub-classes).

I believe `OzoneClient` itself contains a lot of smart logic (especially regarding the buffering and retry/commit) which are good to be tested. This test can be faster and more efficient with real unit tests where the network traffic is mocked.

Fortunately we already have interfaces for both (`OmTransport` and `XceiverClientSpi`. The latter is just an abstract class, not an interface, but nothing can be perfect ;-)).  

With the help of these interfaces it's quite easy to create the unit test.

I copied some methods from the `TestOzoneRpcClientAbstract` and tested them without creating MiniOzoneCluster but using in-memory implementation of OM/DN client interfaces.

Not all the methods are tested (yet) but the main key create / read path is already covered. The main goal is to have all the required classes on master which can be used later by the new EC part of the client which can have good coverage with standard, fast unit tests (and later with integration test, too...)

## How was this patch tested?

The new unit test is executed.

### CODE REVIEW

**Comment by bshashikant** [comment]
@elek , the changes look good. Can you please rebase?

**Comment by bshashikant** [comment]
> @elek , the changes look good so far and i am still reviewing the changes. Meanwhile, Can you please rebase?

**Comment by elek** [comment]
> the changes look good so far and i am still reviewing the changes. Meanwhile, Can you please rebase?

Sure, it's rebased now. (And thank you for checking it).

**Comment by elek** [comment]
Can you please help me with a review @umamaheswararao?

**Review by hanishakoneru — COMMENTED** [review]
Thanks @elek for working on this and building the base suit for OzoneClient testing. We can keep adding more tests on top of it now. 
Patch LGTM. I have one very minor question. Otherwise I am +1 for merge.

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java [hunk 1]
```diff
@@ -248,6 +249,22 @@ public void onRemoval(
         }).build();
   }
 
+  @NotNull
+  @VisibleForTesting
+  protected XceiverClientFactory createXceiverClientFactory(
+      ConfigurationSource configSource,
+      List<X509Certificate> x509Certificates) throws IOException {
+    return new XceiverClientManager(configSource,
+        conf.getObject(XceiverClientManager.ScmClientConfig.class),
```
[the remark this answers is not in the document]
Comment (elek) [reply hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java:258]: Good question (and thanks the review)

`conf` is a private field, so without an additional getter (supposed to be `@VisibleForTesting`) it's not available in the subclasses. That may be my original motivation ;-)

But I just realized that `configSource` is not required in the subclasses when I override the `createXceiverClientFactory`, so I removed it all together:

{code}
     @NotNull
      @Override
      protected XceiverClientFactory createXceiverClientFactory(
          List<X509Certificate> x509Certificates)
          throws IOException {
        return new MockXceiverClientFactory();
      }
{code}

And started to use `conf` field in the main implementation as you suggested.

**Review by hanishakoneru — APPROVED** [review]
Thanks @elek. +1 for merge.

**Comment by elek** [comment]
Thanks for the review @hanishakoneru. Merging it now...

**Pull request merged**